### PR TITLE
feat: adds a multi-region capped local heap cache

### PR DESCRIPTION
**Pull request by jbee**
### Summary
This PR adds a heap memory cache implementation with eager invalidation and memory cap over all caches (here regions).
This implementation is an alternative to the existing `LocalCache` based on cache2k. 
The main purpose is to gain insight about cache usage and see if eager invalidation as implemented can solve the cache problems.

To provide a cap over all local caches the `CappedLocalCache` is not a single cache but a cache cluster containing all local `Cache`s. These are called `CacheRegion` as each implements a single `region`.

To only consume up to a certain size of the available heap memory of the JVM the cache has a cap setting.
Apart from the TTL and default value settings of the cache builder the cache ignores the other builder settings as it has its own strategy to manage cache entries.

To know how much space entries take a `Sizeof` operator was created that **estimates** the size of any given java object in heap memory which is one of the metadata values each cache entry tracks.

The cache performs an assessment of cache entries in an interval. For each entry a number is computed, called burden, which represents how relevant or irrelevant the cache deems the entry. 

* The less read access the more burden. 
* The less TTL left the more burden. 
* The larger the entry size the more burden. 

From this a list of high burden entries is maintained. This list is used if the cache deems it is time to free memory by invalidating entries.

 Reasons to free memory can be: 
* another entry was added and space is nearing the cap (soft cap)
* overall size of cache entries is nearing the cap (hard cap)
* the overall free heap memory is low

### Configuration
A new configuration `ConfigurationKey.SYSTEM_CACHE_CAP_PERCENTAGE` was added that holds the percentage of the heap that all local caches together should use at maximum. Default is `50`%.
If this value is zero or below the capped local cache is disabled and the `LocalCache` based on cache2k is used instead.

### REST API
A controller was added that allows insight into the cache state. This is represented as `CacheInfo` record.
The main purpose of this is to learn what is going on in the caches without having to analyse a heap dump.
The `CacheInfo` record is maintained as apart of the burden list computation that happens in intervals in as a background task.

In addition the RESR API can be used to invalidate all local caches or a particular cache by region.
Also the cap settings of the capped local cache can be changed.

### Automatic Testing
The main implementations of `CappedLocalCache` and the `Sizeof` operator got new unit tests.

### CODE REVIEW

**Review by ameenhere — APPROVED** [review]

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/cache/CacheInfo.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    public static final class CacheGroupInfo
+    {
+        @JsonProperty
+        private final String name;
+
+        @JsonProperty
+        private final int entries;
+
+        @JsonProperty
+        private final long hints;
```
Comment (netroms) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/cache/CacheInfo.java:73]: Typo? hints -> hits?

**Comment by jbee** [comment]
OBS! Before we merge we want to add access protection to the controller and make cache2k the default again.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/cache/CacheInfo.java (hunk 1 above)
Comment (jbee) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/cache/CacheInfo.java:73]: Haha, yes! Good catch.

**Comment by netroms** [comment]
The code style is very compact in some places, a couple of extra newlines would not harm I think :)

**Pull request merged**